### ISSUE TITLE
Alphametics reference solution

### ISSUE THREAD

**Issue by lilislilit** [exercism/python#524]
I noticed, when running test/check-exercises.py that test for alphametics take almost 220 seconds, give or take. I am not sure, maybe there is more optimal solution out there somewhere?

**Comment by behrtam** [comment]
Just looked at this and saw that it is a solution created by me. It looks like a simple brute force approche, so if you can find a better solution that would be great.
I don't remember but I guess I didn't spend much time to finde something that is better/faster.

**Comment by shaleh** [comment]
Wow. It runs in 30s or less on my couple of year old laptop.

**Comment by cmccandless** [comment]
I just looked at the [submissions](http://exercism.io/tracks/python/exercises/alphametics) for this exercise on the track and ran some benchmarks on several of them:

On my system*:

| Solution | Runtime (s) |
| --- | --- |
| [\<reference\>](https://github.com/exercism/python/blob/master/exercises/alphametics/example.py) | 35.04 |
| [@karelvancamp](http://exercism.io/submissions/450df89d4af54cd5a54138b7d1df0cde) | 20.50 |
| [@nikamirrr](http://exercism.io/submissions/8a9baa4086a54a139fbfa34ffae888cc) | 01.06 |

I'm looking into benchmarking the [current submissions](http://exercism.io/tracks/python/exercises/alphametics) for this exercise and while I don't have comprehensive data yet, initial findings show @nikamirrr 's solution to be about 33 times faster than the current reference solution!

@nikamirrr, very well done! Would you be willing to consider letting us include your solution as our reference solution (or even better, submitting a PR yourself updating [the current solution](https://github.com/exercism/python/blob/master/exercises/alphametics/example.py)) ? Currently, `alphametics` is our longest-running test scenario (by far!) in our Travis-CI builds, and it would be fantastic to speed things up.

**Comment by nikamirrr** [comment]
Thank you for the flattering remarks. I remember that problem.
The code I have there may be a bit clumsy.
So, I will see if I can polish it a bit and submit a cleaner version over the weekend.
I did PR once before.

The main trick I used was to screen solution candidates testing the last decimals only and discard bad options early without going after the permutations of the other digits. If you fail, fail as soon as possible :)
Only if the last decimals match, the code proceed to running other digits
